### ISSUE TITLE
Resiliency metrics: add metrics for retry, circuit breaker

### ISSUE THREAD

**Issue by katriendg** [dapr/dapr#7476]
## In what area(s)?

 /area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Describe the feature

Today, based on #4524 which has already been released in the past, when users setup Resiliency configuration policy it is possible to see metrics come through. Based on the doc linked here there are three metrics exposed: https://github.com/dapr/dapr/blob/master/docs/development/dapr-metrics.md#resiliency

* dapr_resiliency_loaded: The number of resiliency policies loaded.
* dapr_resiliency_count: The number of times a resiliency policy has been executed.
* dapr_resiliency_activations_total: Number of times a resiliency policy has been activated in a building block after a failure or after a state change.

This is a useful set of metrics, but when it comes to actual invocations of retries, circuit breakers and timeouts, there is no indication as to the health of the environment. 

Take the example where you have a component like a PubSub component that is experiencing issues with the connection to its broker. Having metrics about retries and circuit breaker kicking in, could help in early detection of a larger problem in the environment. 

The metrics on Resiliency retries, circuit breakers and timeouts would be useful in generating an alert upfront as they start kicking in, while this is unexpected in the system and helps action being taken to address the underlying cause.

## Release Note

RELEASE NOTE: **ADD** New feature in Dapr.

**Comment by alicejgibbons** [comment]
not stale

**Comment by jake-engelberg** [comment]
I've inferred from the description the additional desired resiliency metrics to be:
`dapr_resiliency_cb_current_state`: A resiliency policy's current CircuitBreakerState value.
`dapr_resiliency_retry_activations_total:` Number of times a resiliency policyKey has been activated in a building block after a retry.
`dapr_resiliency_timeout_activations_total`: Number of times a resiliency policyKey has been activated in a building block after a timeout.

Viewing the existing 3 metrics in grafana shows that with some effort, one can determine the desired additional metrics, by using the tags, specifically the `status` and `policyType`. With this in mind, is it still the intention to have additional measures that simplifies the aggregation of the circuit breaker/timeout/retry metrics?
<img width="1199" alt="Screenshot 2024-06-11 at 8 16 22 PM" src="https://github.com/dapr/dapr/assets/152900222/00029887-7146-4c1d-ba9f-b7d4c8a8034c">

Regardless, I have a draft PR of how I general understood the requirements if we were to still go about adding additional measures: https://github.com/dapr/dapr/pull/7807

@yaron2 cc @halspang

**Comment by filintod** [comment]
@katriendg as mentioned above by @jake-engelberg maybe what we need is adding this label information in the docs for clarification? or you have a specific use case that the current metrics don't cover? thx

**Comment by katriendg** [comment]
Hello @filintod @jake-engelberg thanks a lot for looking into this and sharing the details on a way to achieve this with filtering and querying on existing metrics. I'm no longer involved in the project where this came up but it seems this could well cover it. 

Would it make sense to explain this approach in the existing docs? Happy to do a PR in the docs.

**Comment by jake-engelberg** [comment]
> Hello @filintod @jake-engelberg thanks a lot for looking into this and sharing the details on a way to achieve this with filtering and querying on existing metrics. I'm no longer involved in the project where this came up but it seems this could well cover it.
> 
> Would it make sense to explain this approach in the existing docs? Happy to do a PR in the docs.

In terms of retry and timeout counts, those can be queried easily via the `policy` label in the metrics, and a docs PR would suffice:
<img width="254" alt="Screenshot 2024-06-18 at 11 34 16 AM" src="https://github.com/dapr/dapr/assets/152900222/de6f014f-d76c-465b-8341-acbaa7427a08">


As for the circuitbreaker status, I've attempted to make a query that converts the counter (counts circuitbreaker state changes) into a gauge (shows current state of circuitbreaker), but was unsuccessful.

The individual states can be separated in the same way by the `status` label, but I wasn't able to make the query result in showing the current state.
![image](https://github.com/dapr/dapr/assets/152900222/e44d2ffa-6388-49a3-88d4-f6fe74d2aa74)

If there isn't a simple way to do this, it probably is helpful to add a new metric as a gauge counter for the circuit breaker specifically.

**Comment by alicejgibbons** [comment]
Okay so your recommendation @jake-engelberg would be to add a single metric for circuit breaker state and then we could have enough information for the current state of resiliency policies across retries, timeouts and cb?

**Comment by jake-engelberg** [comment]
> Okay so your recommendation @jake-engelberg would be to add a single metric for circuit breaker state and then we could have enough information for the current state of resiliency policies across retries, timeouts and cb?

Yes. While there would be some way to determine the circuit breaker current state by seeing the last incremented series, it doesn't seem possibly to do so easily in a PromQL query, so I suggest adding a gauge metric to do this.

I'll work on the draft here https://github.com/dapr/dapr/pull/7807 to do this single metric and see how it goes. It currently adds 3 metrics and is incomplete.

**Comment by alicejgibbons** [comment]
@jake-engelberg can this be closed?
